spec/models/odk/response_parser_spec.rb

```diff
@@ -0,0 +1,408 @@
+        let(:xml_values) { %w[A B C] }
+        let(:expected_values) { xml_values }
+
+        context "valid input" do
+          it "should produce a simple tree from a form with three children" do
+            Odk::ResponseParser.new(response: response, files: files).populate_response
+
+            expect_children(response.root_node, %w[Answer Answer Answer], form.c.map(&:id), expected_values)
+          end
+        end
+
+        context "with earlier odk code style for questionings" do
+          xit "should parse the response" do
+          end
+        end
+
+        context "outdated form" do
+          let(:xml) { prepare_odk_response_fixture(filename, form, values: xml_values, formver: "wrong") }
+
+          it "should error" do
+            expect do
+              Odk::ResponseParser.new(response: response, files: files).populate_response
+            end.to raise_error(FormVersionError, "Form version is outdated")
+          end
+        end
+
+        context "response contains form item not in form" do
+          let(:other_form) { create(:form) }
+
+          it "should error" do
+            xml # create xml before updating form's second question to have a different form id.
+            form.c[1].update_attribute(:form_id, other_form.id) # skip validations with update_attribute
+            expect do
+              Odk::ResponseParser.new(response: response, files: files).populate_response
+            end.to raise_error(SubmissionError, "Submission contains group or question not found in form.")
+          end
+        end
+      end
+
+      context "with other question types" do
+        let(:xml_values) { ["Quick", "The quick brown fox jumps over the lazy dog", 9.6] }
+        let(:expected_values) { xml_values }
+        let(:question_types) { %w[text long_text decimal] }
+
+        it "processes values correctly" do
+          Odk::ResponseParser.new(response: response, files: files).populate_response
+          expect_children(response.root_node, %w[Answer Answer Answer], form.c.map(&:id), expected_values)
+        end
+      end
+
+      context "response with select multiple" do
+        let(:question_types) { %w[select_multiple select_multiple text] }
+        let(:opt1) { form.c[0].option_set.sorted_children[0] }
+        let(:opt2) { form.c[0].option_set.sorted_children[1] }
+        let(:xml_values) { ["on#{opt1.id} on#{opt2.id}", "none", "A"] }
+        let(:expected_values) { ["#{opt1.option.name};#{opt2.option.name}", nil, "A"] }
+
+        it "should create the appropriate answer tree" do
+          Odk::ResponseParser.new(response: response, files: files).populate_response
+          expect_children(
+            response.root_node,
+            %w[Answer Answer Answer],
+            form.c.map(&:id),
+            expected_values
+          )
+        end
+      end
+
+      # Don't really need this spec; the hard work is in answer.rb and needs test coverage
```

Comment:
I think one to cover integration is ok. Also remember the answer stuff _does_ have coverage.